New: Add external AI modal integration support [ED-20280]

OrTab opened the pull request:
# feat: Add external AI modal integration support

**Ticket:** ED-20280

## Summary

Adds event-driven API for external applications to programmatically trigger Elementor's AI modals, enabling third-party AI integrations and improving developer experience.

## What Changed

### Core Features
- **External Event API**: Listen for `elementor:ai:show-modal` events to trigger AI modals
- **Response Events**: Dispatch `elementor:ai:modal-closed` events with modal results
- **Programmatic Access**: `showAIModal(modalType, location, context, callback)` method
- **Location Navigation**: Direct access to specific AI modal sections (generate, enhance, in-painting)

### Bug Fixes
- **Null Safety**: Fixed context access in `ai-behavior.js` using `getOption()?.controlName`
- **Error Handling**: Added proper error handling for missing context data

## Files Modified
- `modules/ai/assets/js/editor/module.js` - Event handling infrastructure
- `modules/ai/assets/js/editor/ai-behavior.js` - Context access fix & callback support  
- `modules/ai/assets/js/editor/pages/form-media/media-outlet.js` - Location navigation

## Usage Example

```javascript
// Trigger AI modal from external app
window.dispatchEvent(new CustomEvent('elementor:ai:show-modal', {
  detail: {
    modalType: 'media',
    location: 'generate',
    context: { elementId: 'widget-123' }
  }
}));

// Listen for results
window.addEventListener('elementor:ai:modal-closed', (event) => {
  console.log('AI modal completed:', event.detail);
});
```

## Testing
- Manual testing: External event triggering from browser console
- Backward compatibility: Existing AI buttons work unchanged
- Error scenarios: Invalid parameters handled gracefully

## Impact
- Enables external AI tool integrations
- Foundation for AI marketplace features  
- Maintains full backward compatibility
- Improves code safety with null checks

**Resolves:** ED-20280
<!--start_gitstream_placeholder-->
### ✨ PR Description
Purpose: Add external AI modal integration support to allow third-party components to trigger and receive results from Elementor's AI image generation features.

Main changes:
- Implemented custom events system for external modal triggering and result handling
- Added option to hide UI elements for external integrations (back button, AI beta logo)
- Extended AiBehavior with callback support for modal close events and context-aware processing

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using. **[We'd love your feedback!](mailto:product@linearb.io)** 🚀</sub>
<!--end_gitstream_placeholder-->

OrTab committed:
feat: Add external AI modal integration support

- Add event listener for 'elementor:ai:show-modal' events from external integrations
- Implement handleAIModalEvent and showAIModal methods to support programmatic modal opening
- Fix context access in ai-behavior.js using getOption() for null safety
- Add onCloseCallback support for external modal close handling
- Improve media outlet location navigation with explicit location support
- Enable direct navigation to specific AI modal sections (generate, enhance, etc.)

This enables external applications and AI integrations to trigger Elementor's AI modals programmatically while maintaining proper event communication and state management.

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

elementorbot committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

OrTab committed:
feat: enhance AI modal integration with conditional UI hiding

- Hide back button, history icon, and AI beta logo when modals are triggered externally via elementor:ai:show-modal event
- Add isExternalTrigger flag and hideBackButton/withoutHistory/hideAiBetaLogo options to conditionally hide UI elements
- Fix onCloseCallback not being triggered in AiBehavior by properly accessing options via getOption()
- Fix margin-bottom CSS override issue in in-painting-content by using important flag
- Use global settings pattern for hideBackButton to avoid prop drilling through multiple components
- Improve user experience for programmatic AI tool integration

OrTab committed:
feat: set AI modal minimum height to 80vh

- Add minHeight: '80vh' to promptDialogStyleProps for better modal sizing
- Ensures consistent modal height across all AI tools (text, code, media, animation)
- Improves user experience with more spacious interface

OrTab committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

OrTab commented on modules/ai/assets/js/editor/module.js:
```diff
@@ -45,6 +47,9 @@ export default class Module extends elementorModules.editor.utils.Module {
 		elementor.hooks.addFilter( 'controls/base/behaviors', this.registerControlBehavior.bind( this ) );
 		elementor.hooks.addFilter( 'navigator/layout/behaviors', this.registerNavigatorBehavior.bind( this ) );
 
+		// Listen for AI modal events from remote integration
```
Haha so it can be removed no problem 
Its more for sectionalize it

eyalmrejen-elementor commented on modules/ai/assets/js/editor/module.js:
```diff
@@ -251,4 +256,66 @@ export default class Module extends elementorModules.editor.utils.Module {
+				}
+			},
+			setControlValue: ( image ) => {
+				window.dispatchEvent( new CustomEvent( 'elementor:ai:modal-closed', {
```
Since this is the 2nd time this string is used, I would opt to use a constant - avoiding a magic string

eyalmrejen-elementor commented on modules/ai/assets/js/editor/module.js:
```diff
@@ -251,4 +256,66 @@ export default class Module extends elementorModules.editor.utils.Module {
+				window.dispatchEvent( new CustomEvent( 'elementor:ai:modal-closed', {
+					detail: {
+						success: false,
+						error: 'User closed the modal manually without selecting an image',
```
Does this string need to be translated? Or is it just an event log type of thing?
If it does not need, then disregard - just making sure

OrTab replied:
just an event log

eyalmrejen-elementor reviewed (commented):
Missing tests

TzviRabinovitch committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

OrTab committed:
Internal: Extract AI modal event magic string to constant and add tests

- Replace hardcoded 'elementor:ai:modal-closed' string with AI_EVENTS.MODAL_CLOSED constant
- Add comprehensive test coverage for AI modal event dispatching
- Improve code maintainability by centralizing event name definitions

OrTab committed:
Internal: Clean up test comments and finalize AI modal event constant implementation

- Remove unnecessary comments from test file for cleaner code
- Update AI modal event implementation with extracted constants
- Maintain comprehensive test coverage without verbose commenting

OrTab committed:
Internal: Add SHOW_MODAL constant to AI_EVENTS and enhance test coverage

- Extract 'elementor:ai:show-modal' magic string to AI_EVENTS.SHOW_MODAL constant
- Update module.js to use the new constant for event listener
- Add comprehensive test coverage for both modal event constants
- Ensure consistent naming convention for all AI event strings

OrTab commented on tests/jest/unit/modules/ai/assets/js/editor/ai-modal-events.test.js:
```diff
@@ -0,0 +1,237 @@
+		dispatchEventSpy.mockRestore();
+	} );
+
+	describe( 'AI_EVENTS constants', () => {
+		it( 'should have the correct MODAL_CLOSED event name', () => {
+			expect( AI_EVENTS.MODAL_CLOSED ).toBe( 'elementor:ai:modal-closed' );
+		} );
+
+		it( 'should have the correct SHOW_MODAL event name', () => {
+			expect( AI_EVENTS.SHOW_MODAL ).toBe( 'elementor:ai:show-modal' );
+		} );
```
we rely on those events names on our side so if someone changes it the tests will fail

OrTab committed:
Internal: Add note about elementor-ai constants synchronization

- Add comment to remind about keeping AI event constants in sync with elementor-ai
- Ensure consistency between core and AI integration repositories

eyalmrejen-elementor commented on tests/jest/unit/modules/ai/assets/js/editor/ai-modal-events.test.js:
```diff
@@ -0,0 +1,238 @@
+
+			expect( dispatchEventSpy ).toHaveBeenCalledWith(
+				expect.objectContaining( {
+					type: 'elementor:ai:modal-closed',
```
I would test it against the `constant`, and not the value - if you will change the value, you will have to fix the test, and this means that the test is `coupled to the implementation` and not the effect

eyalmrejen-elementor commented on tests/jest/unit/modules/ai/assets/js/editor/ai-modal-events.test.js:
```diff
@@ -0,0 +1,238 @@
+			expect( lastEvent.detail ).toHaveProperty( 'success', true );
+			expect( lastEvent.detail ).toHaveProperty( 'modalType', 'test' );
+			expect( lastEvent.detail ).toHaveProperty( 'location', 'test-location' );
+			expect( lastEvent.detail ).toHaveProperty( 'id', 789 );
```
I don't know if this is a question of style or if you like to type/copy paste
I would refactor it to:
```
expect( lastEvent.detail ).toHaveProperty( 'id', testImage.id );
expect( lastEvent.detail ).toHaveProperty( 'url', testImage.,url);
```
Now the test is no longer coupled to the actual values

eyalmrejen-elementor reviewed (dismissed):
Just look at the comments - we can continue with this PR, but next time, I think we can better fine grane the test, and have them less coupled to the implementaion and more towards the test intent.

OrTab committed:
refactor: improve AI modal events test practices

- Remove hardcoded values and use actual test data
- Use constants instead of string literals for event types
- Remove configuration tests that only verify constant values
- Focus tests on behavior rather than implementation details
- Fix linting issues and improve maintainability

Following code review feedback to make tests less coupled to implementation and more maintainable.

OrTab committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

MichaelLiamin committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

OrTab committed:
Merge branch 'main' into feat/ai-modal-integration-improvements

OrTab committed:
Empty commit to trigger CI/CD

OrTab committed:
Merge branch 'main' into feat/ai-modal-integration-improvements
